ngraph_bridge/ngraph_executor.cc

```diff
@@ -240,190 +219,169 @@ Status NGraphExecutor::GetNgExecutable(
     return errors::Internal("Backend not available: ", m_op_backend_name);
   }
 
-  lock_guard<mutex> lock(m_mutex);
-
-  std::stringstream signature_ss;
-  std::vector<TensorShape> input_shapes;
-  std::vector<const Tensor*> static_input_map;
-  TF_RETURN_IF_ERROR(ComputeSignature(tf_input_tensors, input_shapes,
-                                      static_input_map, signature_ss));
-  string signature;
-  signature = signature_ss.str();
-
-  NGRAPH_VLOG(5) << "Computed signature: " << signature;
-
-  cache_hit = false;
-  auto it = m_ng_exec_map.find(signature);
-
-  // Translate the TensorFlow graph to nGraph.
-  if (it == m_ng_exec_map.end()) {
-    // Measure the current total memory usage
-    long vm, rss, vm0, rss0;
-    MemoryProfile(vm0, rss0);
-
-    NGRAPH_VLOG(1) << "Compilation cache miss: " << m_node_name;
-
-    std::shared_ptr<ngraph::runtime::Executable> evicted_ng_exec;
-    string serialized_ng_func;
-    std::shared_ptr<ngraph::Function> ng_function;
-    if (!m_do_aot) {
-      TF_RETURN_IF_ERROR(Builder::TranslateGraph(input_shapes, static_input_map,
-                                                 m_graph.get(), ng_function));
-      ng_function->set_friendly_name(m_node_name);
-      int json_indentation = 4;
-      serialized_ng_func = ngraph::serialize(ng_function, json_indentation);
-    } else {
-      auto itr = m_aot_functions.find(signature);
-      if (itr == m_aot_functions.end()) {
-        return errors::Internal(
-            "Expected to find AOT precompiled ng function of signature: ",
-            signature);
-      }
-      serialized_ng_func = itr->second;
-    }
+  cache_hit = true;
+  // Generate forwarding call to Callback functions
+  // CreateCallback and DestroyCallback
+  auto create_ng_items_callback =
+      std::bind(&NGraphExecutor::CreateCallback, this, std::placeholders::_1,
+                input_shapes, static_input_map, op_backend, &cache_hit);
+  auto destroy_ng_items_callback =
+      std::bind(&NGraphExecutor::DestroyCallback, this, std::placeholders::_1,
+                op_backend);
+  // Get NgItems i.e. ng_executable, serialized ng_functions from Data Cache
+  auto status_ng_item_pair = m_ng_data_cache.LookUpOrCreate(
+      signature, create_ng_items_callback, destroy_ng_items_callback);
+
+  if (status_ng_item_pair.first == Status::OK()) {
+    ng_exec = std::get<0>(status_ng_item_pair.second);
+    pts = std::get<2>(status_ng_item_pair.second);
+  }
```

Comment:
get<1> gives the function, I guess? We should be returning that as well. And using that to try-catch the `ng_exec->call` and dump the serialized function in case of errors: https://github.com/tensorflow/ngraph-bridge/blob/vishakh1/MergingtoExecutor/ngraph_bridge/ngraph_encapsulate_op.cc#L459

currently the `ng_exec->call` is not try-catched in master, but maybe we can use this opportunity to add back the try-catch, similar to the `impl` case: https://github.com/tensorflow/ngraph-bridge/blob/master/ngraph_bridge/ngraph_encapsulate_op.cc#L761